src/Covenant/Analysis/Poetry/PoetryLicenseParser.cs

```diff
@@ -0,0 +1,120 @@
+using System.Text.RegularExpressions;
+
+namespace Covenant.Analysis.Poetry;
+
+internal static class PoetryLicenseParser
+{
+    public static BomLicense? Parse(string sitePackagesPath, PoetryLockPackage package, AnalysisContext context)
+    {
+        var packageName = package.Name!.Replace("-", "_");
```

Comment:
This is another example of how the nullability looks like it's not quite right. Clearly this code just expects `Name` to be non-null, which in turn implies that `Name` should not have a nullable type.

The only exception would be if there are sometimes cases where `Name` legitimately is `null`, but you happen to know that this particular code is only ever invoked in cases where `Name` happens not to be `null`.

More generally, the use of `!` is necessary in cases where context-specific knowledge means that you can count on something being non-null but there are other contexts in which it might be null. These are vexing but sometimes unavoidable. But if that's not your situation, `!` is a code smell.